os/fs/smartfs/smartfs_utils.c

```diff
@@ -1220,51 +1290,18 @@ int smartfs_finddirentry(struct smartfs_mountpt_s *fs, struct smartfs_entry_s *d
 							strncpy(direntry->name, entry->name, fs->fs_llformat.namesize);
 							direntry->datalen = 0;
 
-							/* Scan the file's sectors to calculate the length and perform
-							 * a rudimentary check.
+							/* Invoke smartfs_get_datalen to scan through the file's sectors to calculate
+							 * its length and perform a rudimentary check.
 							 */
-
 #ifdef CONFIG_SMARTFS_ALIGNED_ACCESS
 							if ((smartfs_rdle16(&entry->flags) & SMARTFS_DIRENT_TYPE) == SMARTFS_DIRENT_TYPE_FILE) {
 								dirsector = smartfs_rdle16(&entry->firstsector);
 #else
 							if ((entry->flags & SMARTFS_DIRENT_TYPE) == SMARTFS_DIRENT_TYPE_FILE) {
 								dirsector = entry->firstsector;
 #endif
-								while (dirsector != SMARTFS_ERASEDSTATE_16BIT) {
-									/* Read the next sector of the file */
-
-									smartfs_setbuffer(&readwrite, dirsector, 0, sizeof(struct smartfs_chain_header_s), (uint8_t *)fs->fs_rwbuffer);
-									ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
-									if (ret < 0) {
-										fdbg("Error in sector chain at %d, ret : %d\n", readwrite.logsector, ret);
-										break;
-									}
-#ifdef CONFIG_SMARTFS_DYNAMIC_HEADER
-									if (SMARTFS_NEXTSECTOR(header) == SMARTFS_ERASEDSTATE_16BIT) {
-
-										readwrite.count = fs->fs_llformat.availbytes;
-										readwrite.buffer = (uint8_t *)fs->fs_chunk_buffer;
-
-										ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
-										if (ret < 0) {
-											fdbg("Error reading sector %d header, ret : %d\n", readwrite.logsector, ret);
-											break;
-										}
-										used_value = get_leftover_used_byte_count((uint8_t *)readwrite.buffer, get_used_byte_count((uint8_t *)header->used));
-										direntry->datalen += used_value;
-									} else {
-										direntry->datalen += SMARTFS_AVAIL_DATABYTES(fs);
-									}
-									readwrite.buffer = (uint8_t *)fs->fs_rwbuffer;
-#else
-									/* Add used bytes to the total and point to next sector */
-									if (SMARTFS_USED(header) != SMARTFS_ERASEDSTATE_16BIT) {
-										direntry->datalen += SMARTFS_USED(header);
-									}
-#endif
-									dirsector = SMARTFS_NEXTSECTOR(header);
-								}
+								/* Mark the file's length as unknown, it will be calculated later if required */
+								direntry->datalen = SMARTFS_DIRENT_LEN_UNKWN;
```

Comment:
this datalen will be used only for file but it should initialized for both of file & directory and very early time. you should not fix it in this time, but later.

Reply:
For files, datalen is first initialized as 0, and then changed to DATALEN_UNKWN 
This us so that there no confusion between file length 0 and file length not calculated

For directories, datalen is 0, because DIRENT_TYPE helps identify the entry as a directory, so data length has no meaning